dnscache.go

```diff
@@ -246,17 +246,45 @@ func (r *Resolver) storeLocked(key string, rrs []string, used bool, err error) {
 	}
 }
 
-var defaultResolver = &defaultResolverWithTrace{}
+var defaultResolver = &defaultResolverWithTrace{
+	ipVersion: "ip",
+}
+
+// Create a new resolver that only resolves to IPv4 Addresses when looking up Hosts.
+// Example:
+//
+//	resolver := dnscache.Resolver{
+//	  Resolver: NewResolverOnlyV4(),
+//	}
+func NewResolverOnlyV4() DNSResolver {
+	return &defaultResolverWithTrace{
+		ipVersion: "ip4",
+	}
+}
+
+// Create a new resolver that only resolves to IPv6 Addresses when looking up Hosts.
+// Example:
+//
+//	resolver := dnscache.Resolver{
+//	  Resolver: NewResolverOnlyV6(),
+//	}
+func NewResolverOnlyV6() DNSResolver {
+	return &defaultResolverWithTrace{
+		ipVersion: "ip6",
+	}
+}
 
 // defaultResolverWithTrace calls `LookupIP` instead of `LookupHost` on `net.DefaultResolver` in order to cause invocation of the `DNSStart`
 // and `DNSDone` hooks. By implementing `DNSResolver`, backward compatibility can be ensured.
-type defaultResolverWithTrace struct{}
+type defaultResolverWithTrace struct {
+	ipVersion string
+}
 
 func (d *defaultResolverWithTrace) LookupHost(ctx context.Context, host string) (addrs []string, err error) {
 	// `net.Resolver#LookupHost` does not cause invocation of `net.Resolver#lookupIPAddr`, therefore the `DNSStart` and `DNSDone` tracing hooks
 	// built into the stdlib are never called. `LookupIP`, despite it's name, can also be used to lookup a hostname but does cause these hooks to be
 	// triggered. The format of the reponse is different, therefore it needs this thin wrapper converting it.
-	rawIPs, err := net.DefaultResolver.LookupIP(ctx, "ip", host)
+	rawIPs, err := net.DefaultResolver.LookupIP(ctx, d.ipVersion, host)
```

Comment:
You need to handle the case where d.ipVersion is "".

Reply:
I would argue that this is not necessary as the only way to use this `defaultResolverWithTrace` (from outside this lib) is by either using the `defaultResolver` or by retreiving a `defaultResolverWithTrace` from the functions `NewResolverOnlyV4` and `NewResolverOnlyV6`. In all 3 cases there is alreade the ipVersion set.

Reply:
Sure but code evolve over time and this situation could change.

Reply:
ok i've added that if ipVersion is not ip, ip4 or ip6 then ip will be used